### PR TITLE
[RFC] introduce ATF support on NXP ls1043a platform

### PR DESCRIPTION
**Pull request by JiafeiPan**
The LS1043A QorIQ advanced multicore processor combines two to four ARM
Cortex®-v8 A53 cores with datapath acceleration optimized for L2/3 packetc processing, single pass security offload, and robust traffic management and quality of service. More information can be found at NXP official website.

This patchset includes the following feature on ls1043ardb:
1. PSCI support.
2. Support ATF boot from Nor flash and SD Card.
3. ATF boot bl33(uboot or uefi).
4. ATF boot Tee OS on ls1043a platform.

Please help to view these patches, any comments is welcome, thanks.

### CODE REVIEW

File: plat/layerscape/common/ls_io_storage.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	.length = PLAT_LS_FIP_MAX_SIZE
+};
+
+static const io_uuid_spec_t bl2_uuid_spec = {
+	.uuid = UUID_TRUSTED_BOOT_FIRMWARE_BL2,
+};
+
+static const io_uuid_spec_t scp_bl2_uuid_spec = {
+	.uuid = UUID_SCP_FIRMWARE_SCP_BL2,
+};
```
Comment (ruchi393) [inline plat/layerscape/common/ls_io_storage.c]: SCP Image is not present on Layerscape platform , so this identifier is not required

File: plat/layerscape/common/ls_io_storage.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	[BL2_IMAGE_ID] = {
+		&fip_dev_handle,
+		(uintptr_t)&bl2_uuid_spec,
+		open_fip
+	},
+	[SCP_BL2_IMAGE_ID] = {
+		&fip_dev_handle,
+		(uintptr_t)&scp_bl2_uuid_spec,
+		open_fip
+	},
```
Comment (ruchi393) [inline plat/layerscape/common/ls_io_storage.c]: SCP reference should be removed from here

File: plat/layerscape/common/ls_common.mk [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+PLAT_BL_COMMON_SOURCES	+=	${XLAT_TABLES_LIB_SRCS}
+endif
+
+BL1_SOURCES		+=			\
+				drivers/io/io_fip.c				\
+				drivers/io/io_memmap.c				\
+				drivers/io/io_storage.c				\
+				plat/layerscape/common/ls_timer.c			\
+				plat/layerscape/common/ls_bl1_setup.c			\
+				plat/layerscape/common/ls_io_storage.c
```
Comment (ruchi393) [inline plat/layerscape/common/ls_common.mk]: BL1 is not required on layerscape platforms.

The flow should be BL2->Bl31

You should use BL2_AT_EL3 option for this.

File: docs/plat/ls1043a.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+More information are listed in `ls1043`_.
+
+Boot Sequence
+=============
+
+
+Bootrom --> ATF BL1 --> ATF BL2 --> ATF BL1 --> ATF BL31
+--> BL32(Tee OS) --> ATF BL31 --> BL33(u-boot) --> Linux kernel
+
```
Comment (ruchi393) [inline docs/plat/ls1043a.rst]: The flow needs to be changes to  -

Bootrom -> ATF BL2-> ATF Bl31 -> BL32 ->Bl31 -> BL33

Additional BL1 step is not required on Layerscape platforms

File: plat/layerscape/common/ls_io_storage.c (hunk 1 above)
Comment (JiafeiPan) [reply plat/layerscape/common/ls_io_storage.c]: yes, will remove it, thanks

File: plat/layerscape/common/ls_io_storage.c (hunk 2 above)
Comment (JiafeiPan) [reply plat/layerscape/common/ls_io_storage.c]: Will remove, thanks.

File: plat/layerscape/common/ls_common.mk (hunk 3 above)
Comment (JiafeiPan) [reply plat/layerscape/common/ls_common.mk]: In this design, BL2 runs in EL1, so we must use BL1.  Maybe BL2 runs in EL3 is another option, it is not implemented in this patch.

File: docs/plat/ls1043a.rst (hunk 4 above)
Comment (JiafeiPan) [reply docs/plat/ls1043a.rst]: In this design, BL2 runs in EL1, so we must use BL1.  Maybe BL2 runs in EL3 is another option, it is not implemented in this patch.

**Comment by JiafeiPan** [comment]
I will create formal pull request, thanks

**Pull request closed without merging**